File: 1160-find-words-that-can-be-formed-by-characters/1160-find-words-that-can-be-formed-by-characters.cpp
class Solution {
public:
    int countCharacters(vector<string>& words, string chars) {
        map<char, int> charBank;
        for (char ch : chars) {
            if (charBank[ch]) {
                charBank[ch] += 1;
            } else {
                charBank[ch] = 1;
            }
        }
        int result = 0;
        map<char, int> copyBank;
        for (string word : words) {
            copyBank = charBank;
            if (validateAnagram(copyBank, word)) {
                result += word.size();
            }
            copyBank.clear();
        }
        return result;
    }
    bool validateAnagram(map<char, int> obj, string str) {
        for (char ch : str) {
            if (obj[ch]) {
                obj[ch] -= 1;
            } else {
                return false;
            }
        }
        return true;
    }
};

File: 1160-find-words-that-can-be-formed-by-characters/1160-find-words-that-can-be-formed-by-characters.js
/**
 * @param {string[]} words
 * @param {string} chars
 * @return {number}
 */
var countCharacters = function(words, chars) {
    let charBank = {};
    for (ch of chars) {
        if (charBank[ch]) {
            charBank[ch] += 1;
        } else {
            charBank[ch] = 1;
        }
    }
    let result = 0;
    for (let word of words) {
        let copyBank = Object.assign({}, charBank);
        if (validateAnagram(copyBank, word)) {
            result += word.length;
        }
    }
    return result;
};

function validateAnagram(obj, str) {
    for (let ch of str) {
        if (obj[ch]) {
            obj[ch] -= 1;
        } else {
            return false;
        }
    }
    return true;
}